File: backend/src/api/services/graph.py
import asyncio
import networkx as nx

import api.services as services
import api.utils.nodes as nodes


class ComputeGraph(nx.DiGraph):
    @property
    def elements(self):
        for id in self.nodes:
            yield self._node_to_dict(id)
        for u, v in self.edges:
            yield from self._edge_to_list(u, v)

    @staticmethod
    def _broadcast_update(func):
        async def wrapper(*args, **kwargs):
            if type(res := func(*args, **kwargs)) is list:
                await asyncio.gather(
                    *[services.websocket_handler.broadcast("graph", msg) for msg in res]
                )
            else:
                await services.websocket_handler.broadcast("graph", res)

        return wrapper

    @_broadcast_update
    def add_node(self, id: int, type: str, params: dict, pos: tuple[int]) -> dict:
        obj = nodes.constructors[type](params, pos)

        super().add_node(id, obj=obj)

        return {"action": "create_node", "node": self._node_to_dict(id)}

    @_broadcast_update
    def update_node(
        self, id: int, params: dict | None = None, pos: tuple[int] | None = None
    ) -> dict:
        obj: nodes.Node = self.nodes[id]["obj"]

        msg = {"action": "update_node", "node": {"id": id}}

        if params:
            obj.params = params
            msg["node"].update({"data": {"values": params}})
        if pos:
            obj.pos = pos
            msg["node"].update({"position": {"x": pos[0], "y": pos[1]}})

        return msg

    @_broadcast_update
    def remove_node(self, id: int) -> dict:
        super().remove_node(id)

        return {"action": "remove_node", "id": str(id)}

    @_broadcast_update
    def add_edge(self, u: int, v: int, map_: dict) -> dict:
        super().add_edge(u, v, map=map_)

        return [{"action": "create_edge", "edge": e} for e in self._edge_to_list(u, v)]

    @_broadcast_update
    def update_edge(self, u: int, v: int, map_: dict) -> dict:
        actions = [
            {"action": "remove_edge", "id": f"e{u}{uh}-{v}{vh}"}
            for uh, vh in self.edges[u, v]["map"]
        ]

        self.edges[u, v]["map"] = map_

        actions += [
            {"action": "create_edge", "edge": e} for e in self._edge_to_list(u, v)
        ]

        return actions

    @_broadcast_update
    def remove_edge(self, u: int, v: int) -> dict:
        map_ = self.edges[u, v]["map"]
        super().remove_edge()
        return [{"action": "remove_edge", "id": f"e{u}{uh}-{v}{vh}"} for uh, vh in map_]

    def _node_to_dict(self, id: int) -> dict:
        obj: nodes.Node = self.nodes[id]["obj"]

        return {
            "id": str(id),
            "label": (name := type(obj).__name__),
            "type": "node",
            "data": {"template": obj.template_computed, "values": obj.params},
            "position": {"x": obj.pos[0], "y": obj.pos[1]},
        }

    def _edge_to_list(self, u: int, v: int) -> list[dict]:
        map_: dict[str, str] = self.edges[u, v]["map"]

        return [
            {
                "id": f"e{u}{uh}-{v}{vh}",
                "source": u,
                "sourceHandle": uh,
                "target": v,
                "targetHandle": vh,
            }
            for uh, vh in map_.items()
        ]


compute_graph = ComputeGraph()

File: frontend/src/stores/graph.ts
import { defineStore } from 'pinia'

import { ref } from 'vue'
import type { Ref } from 'vue'

import type { Node, Edge, Connection } from '@vue-flow/core'

import { app } from '@/main'
import { webSocketHandler } from '@/services/websocket'
import { useFilesStore } from '@/stores/files'
import type { Directory } from '@/stores/files'

export interface Template {
  inputs: {
    [key: string]: {
      name: string
      type: string
    }
  }
  outputs: {
    [key: string]: {
      name: string
      type: string
    }
  }
  values: {
    [key: string]: {
      name: string
      component: { type: string; [key: string]: any }
    }
  }
}

interface NodeSchema {
  id: number
  type: string
  values: { [key: string]: any }
  position: { x: number; y: number }
}

interface EdgeSchema {
  id: string
  source: number
  sourceHandle: string
  target: number
  targetHandle: string
}

export const useGraphStore = defineStore('graph', () => {
  const filesStore = useFilesStore()

  const templates: Ref<{ [key: string]: Template }> = ref({})
  const nodes: Ref<Node[]> = ref([])
  const edges: Ref<Edge[]> = ref([])
  const version = ref(0)

  function getNode(id: number) {
    return nodes.value.find((node) => node.id === `${id}`)
  }

  function startListening() {
    webSocketHandler.send('streams', { action: 'subscribe', streams: ['graph'] })

    fetchGraph()
  }

  async function fetchGraph() {
    const {
      version: version_,
      templates: templates_,
      nodes: nodes_,
      edges: edges_
    } = await fetch(new URL('graph/', app.config.globalProperties.apiURL), {
      method: 'GET',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json;charset=UTF-8'
      }
    }).then((res) => {
      return res.json()
    })

    version.value = version_
    templates.value = templates_

    nodes.value = nodes_.map(nodeToVueFlow)
    edges.value = edges_.map(edgeToVueFlow)
  }

  function nodeToVueFlow(node: NodeSchema): Node {
    return {
      id: `${node.id}`,
      type: 'node',
      position: node.position,
      data: { type: node.type, values: node.values },
      isValidTargetPos: connectionValid,
      isValidSourcePos: connectionValid
    }
  }

  function edgeToVueFlow(edge: EdgeSchema): Edge {
    return {
      id: edge.id,
      type: 'edge',
      updatable: true,
      source: `${edge.source}`,
      sourceHandle: edge.sourceHandle,
      target: `${edge.target}`,
      targetHandle: edge.targetHandle
    }
  }

  webSocketHandler.addEventListener('message', (event) => {
    const msg = (<CustomEvent>event).detail

    if (msg.stream != 'graph') return

    const data = msg.data as {
      version: number
      action: string
      id?: string | number
      node?: NodeSchema
      edge?: EdgeSchema
    }

    if (data.version <= version.value) return
    else if (data.version > version.value + 1) {
      fetchGraph()
      return
    }
    version.value++

    switch (data.action) {
      case undefined:
        throw new Error(`Required value 'action' not received`)
      case 'createNode':
        if (!data.node) throw new Error(`Required value 'node' not received`)

        nodes.value.push(nodeToVueFlow(data.node))
        break
      case 'deleteNode':
        if (!data.id) throw new Error(`Required value 'id' not received`)
        if (typeof data.id === 'string') throw new Error(`'id' is of type 'string'`)

        let idx = nodes.value.findIndex((node) => node.id === `${data.id}`)

        if (idx > -1) nodes.value.splice(idx, 1)
        else throw new Error(`Invalid node ID ${data.id}`)

        break
      case 'updatePositionNode':
        if (!data.node) throw new Error(`Required value 'node' not received`)

        const node = getNode(data.node.id)

        if (!node) {
          fetchGraph()
          throw new Error(`Node '${data.node.id}' does not exist, resyncing graph...`)
        }

        node.position = data.node.position

        break
      case 'updateValuesNode':
        if (!data.node) throw new Error(`Required value 'node' not received`)

        const node_ = getNode(data.node.id)

        if (!node_) {
          fetchGraph()
          throw new Error(`Node '${data.node.id}' does not exist, resyncing graph...`)
        }

        node_.data.values = { ...node_.data.values, ...data.node.values }

        break

      case 'createEdge':
        if (!data.edge) throw new Error(`Required value 'edge' not received`)

        edges.value.push(edgeToVueFlow(data.edge))

        break
      case 'deleteEdge':
        if (!data.id) throw new Error(`Required value 'id' not received`)
        if (typeof data.id === 'number') throw new Error(`'id' is of type 'number'`)

        idx = edges.value.findIndex((edge) => edge.id === data.id)

        if (idx > -1) edges.value.splice(idx, 1)
        else throw new Error(`Invalid edge ID ${data.id}`)

        break
      default:
        throw new Error(`Unrecognized action '${data.action}''`)
    }
  })

  if (webSocketHandler.active) startListening()

  webSocketHandler.addEventListener('open', startListening)

  function addNode(type: string, position: { x: number; y: number }) {
    const values: { [key: string]: any } = {}

    for (const [k, v] of Object.entries(templates.value[type].values)) {
      if (v.component.type === 'FileDropdown') {
        function recurse(subStructure: Directory): string[] {
          let [dir, file]: [string | null, Directory | null] = [null, null]
          for ([dir, file] of Object.entries(subStructure)) if (!file) return [dir]

          if (!dir || !file) return ['']

          return [dir, ...recurse(file)]
        }

        values[k] = recurse(filesStore.getSubStructure(v.component.directory))
      } else {
        values[k] = v.component.default
      }
    }

    const node: NodeSchema = {
      id: version.value,
      type: type,
      values: values,
      position: position
    }

    nodes.value.push(nodeToVueFlow(node))

    version.value++

    console.log({
      item: {
        version: version.value - 1,
        action: 'createNode',
        ...node
      }
    })

    webSocketHandler.send('graph', {
      item: {
        version: version.value - 1,
        action: 'createNode',
        ...node
      }
    })

    return node.id
  }

  function removeNode(id: number) {
    const idx = nodes.value.findIndex((node) => node.id === `${id}`)

    if (idx > -1) nodes.value.splice(idx, 1)
    else throw new Error(`Invalid node ID ${id}`)

    version.value++

    webSocketHandler.send('graph', {
      item: {
        version: version.value - 1,
        action: 'deleteNode',
        id: id
      }
    })
  }

  function updatePositionNode(id: number, position: { x: number; y: number }) {
    const node = nodes.value.find((node) => node.id === `${id}`)

    if (!node) throw new Error(`Invalid node ID '${id}'`)

    version.value++

    node.position = position

    webSocketHandler.send('graph', {
      item: {
        version: version.value - 1,
        action: 'updatePositionNode',
        id: id,
        position: position
      }
    })
  }

  function updateValuesNode(id: number, values: { [key: string]: any }) {
    const node = nodes.value.find((node) => node.id === `${id}`)

    if (!node) throw new Error(`Invalid node ID '${id}'`)

    version.value++

    node.data.values = { ...node.data.values, ...values }

    webSocketHandler.send('graph', {
      item: {
        version: version.value - 1,
        action: 'updateValuesNode',
        id: id,
        values: values
      }
    })
  }

  function addEdge(source: number, target: number, sourceHandle: string, targetHandle: string) {
    const edge: EdgeSchema = {
      id: `e${source}${sourceHandle}-${target}${targetHandle}`,
      source: source,
      target: target,
      sourceHandle: sourceHandle,
      targetHandle: targetHandle
    }

    version.value++

    edges.value.push(edgeToVueFlow(edge))

    webSocketHandler.send('graph', {
      item: {
        version: version.value - 1,
        action: 'createEdge',
        ...edge
      }
    })

    return edge.id
  }

  function removeEdge(id: string) {
    const idx = edges.value.findIndex((edge) => edge.id === id)

    if (idx > -1) edges.value.splice(idx, 1)
    else throw new Error(`Invalid edge ID ${id}`)

    version.value++

    webSocketHandler.send('graph', {
      item: {
        version: version.value - 1,
        action: 'deleteEdge',
        id: id
      }
    })
  }

  function connectionValid(connection: Connection) {
    const source = nodes.value.find((node) => node.id === connection.source)
    const target = nodes.value.find((node) => node.id === connection.target)

    if (!source) throw new Error(`Invalid source node '${connection.source}'`)
    if (!target) throw new Error(`Invalid target node '${connection.target}'`)

    if (!connection.sourceHandle) throw new Error(`Connection requires 'sourceHandle'`)
    if (!connection.targetHandle) throw new Error(`Connection requires 'targetHandle'`)

    const sourceType = templates.value[source.data.type].outputs[connection.sourceHandle].type

    const targetType = templates.value[target.data.type].inputs[connection.targetHandle].type

    return sourceType === targetType
  }

  return {
    nodes,
    edges,
    version,
    templates,
    connectionValid,
    addNode,
    removeNode,
    updatePositionNode,
    updateValuesNode,
    addEdge,
    removeEdge
  }
})
